### PR TITLE
[ONNX] Apply modularization pass to exported program exporting

### PR DESCRIPTION
**Pull request by titaiwangms**
Motivation
---

Modularization provides users better experience on debugging their model by remaining module hierarchy, and it's already applied to nn module exporting, but on exported program, it was segment fault when we applied it at the first. With the development of aot and exported program, now it doesn't segment fault any more.

Features
---

1. Exported program is now exported the same graph as the one from torch.nn.Module, and it potentially provides higher success rate because aot autograd addresses inplace ops.
2. The same graph promise also ensures the development on onnx-rewriter won't be affected.
3. To support fake mode with non-persistent buffer and constant tensors, adapter accesses the value by ExportedProgram.constants. (Need follow-ups)

Methods
---

There are two situations in the current exporter: (1) real model exporting and (2) fake mode. The idea of graph surgery is the same, which exporter manages to stuff lifted parameters, buffers, and constant tensors back to the model by leveraging on `nn_module_stack` of each node. The steps include sink `placeholder`, and create `get_attr` node with assiging real value from exported program providing stat_dict or constants property.

However, in fake mode exporting, if it's exported within fake mode, the non-persistent buffers would be fakified, while constant tensors would not, and different from the normal parameters and buffers, they are not accessible to state_dict(). Without a formal API, such as torch.save for state_dict, we currently, use input adapter to provide it back to model inputs.

Follow-ups
---

For constant tensors and non-persistent buffers, they are not part of state_dict(). Thus, we are using `adapt_torch_inputs` API to stuff them back from ExportedProgram.constants. However, this is a hack. To address this, the followers should explore what we can get from huggingface .bin files to see whether non-persistent buffers and constant tensors is actually accessible.

~~1. Should we keep non-persistent buffers and constants lifted? It's observed that in current GraphModule that the same attribute could be used by multiple `get_attr` nodes.~~
2. Remove the hack that extends `ExportedProgram.constants` to ONNX inputs (https://github.com/pytorch/pytorch/issues/115461)

### CODE REVIEW

File: torch/onnx/_internal/io_adapter.py [hunk 1]
```diff
@@ -67,9 +66,8 @@ def append_step(self, step: InputAdaptStep) -> None:
     def apply(
         self,
         *model_args,
-        model: Optional[
-            Union[torch.nn.Module, Callable, torch_export.ExportedProgram]
-        ] = None,
+        model: Optional[torch_export.ExportedProgram] = None,
```
Comment (BowenBao) [inline torch/onnx/_internal/io_adapter.py:69]: is the type annotation change intended?

File: torch/onnx/_internal/io_adapter.py [hunk 2]
```diff
@@ -131,21 +130,18 @@ def append_step(self, step: OutputAdaptStep) -> None:
...
     ) -> Sequence[Union["torch.Tensor", int, float, bool, str]]:
         """Converts the PyTorch model outputs to exported ONNX model outputs format.
 
         Args:
             model_outputs: The PyTorch model outputs.
-            model: The PyTorch model.
```
Comment (BowenBao) [inline torch/onnx/_internal/io_adapter.py:142]: the arg still exist

File: torch/onnx/_internal/io_adapter.py (hunk 1 above)
Comment (titaiwangms) [reply torch/onnx/_internal/io_adapter.py:69]: Yes, I changed the interface of `adapt_torch_inputs_XX`. It was Union to compromise model=None situation, and now I added some if/else to detach the unrelated model type.

File: torch/onnx/_internal/io_adapter.py [hunk 3]
```diff
@@ -580,11 +571,12 @@ def apply(
         return flattened_outputs
 
 
-class PrependParamsBuffersConstantAotAutogradInputStep(InputAdaptStep):
-    """Prepend model parameters, buffers and constants to the user input.
+class AppendLiftedConstantNonPersistentBufferInputStep(InputAdaptStep):
+    """Prepend model lifted constants to the user input.
```
Comment (BowenBao) [inline torch/onnx/_internal/io_adapter.py:575]: name says Append comment says Prepend.

File: torch/onnx/_internal/exporter.py [hunk 4]
```diff
@@ -882,29 +895,38 @@ def adapt_torch_inputs_to_onnx(
...
+        if (
+            "onnx_input_names" in model_kwargs
+            and model_kwargs["onnx_input_names"] is None
+        ):
+            onnx_input_names = model_kwargs.pop("onnx_input_names")
+
         return self._input_adapter.apply(
-            *model_args, model=model_with_state_dict, **model_kwargs
+            *model_args,
+            model=exported_program,
```
Comment (BowenBao) [inline torch/onnx/_internal/exporter.py:920]: Is this the place you mentioned in https://github.com/pytorch/pytorch/pull/119211/files#r1480511928, where as the model argument is not passed in for other type of graph extractors?

File: torch/onnx/_internal/exporter.py [hunk 5]
```diff
@@ -882,29 +895,38 @@ def adapt_torch_inputs_to_onnx(
...
         """
-        # model specified by the user has precedence, when specified
-        model_with_state_dict = model_with_state_dict or self._model_torch
-        assert (
-            model_with_state_dict is not None
-        ), "model_with_state_dict must be specified."
+        # backward compatibility: Respect the usage of *args, **kwargs for users and
+        # also None for the optional arguments
+        if (
+            "exportped_program" in model_kwargs
```
Comment (BowenBao) [inline torch/onnx/_internal/exporter.py:908]: I don't think it's possible for the below two cases. Plus there is a typo XD

Comment (titaiwangms) [reply torch/onnx/_internal/exporter.py:908]: Huge mistake

File: torch/onnx/_internal/exporter.py [hunk 6]
```diff
@@ -921,9 +943,7 @@ def adapt_torch_outputs_to_onnx(
 
         Args:
             model_outputs: The PyTorch model outputs.
-            model_with_state_dict: The PyTorch model to get extra state from.
-                If not specified, the model used during export is used.
-                Required when :func:`enable_fake_mode` is used to extract real initializers as needed by the ONNX graph.
+            model: The Pytorch model
```
Comment (BowenBao) [inline torch/onnx/_internal/exporter.py:946]: Suggest documenting motivation of why this argument is needed.

I wasn't sure even based on removed comment. Since this is dealing with output instead of input.

File: test/onnx/test_fx_to_onnx_with_onnxruntime.py [hunk 7]
```diff
@@ -1067,23 +1067,48 @@ def _test_fake_tensor_mode_exporter(
...
+                if (
+                    model_type
+                    == pytorch_test_common.TorchModelType.TORCH_EXPORT_EXPORTEDPROGRAM
+                ):
+                    input_names, _ = onnx_test_common.get_ort_input_names_and_session(
+                        tmp_onnx_file.name
+                    )
+                    args_not_none = onnx_program.adapt_torch_inputs_to_onnx(
+                        *args,
+                        exported_program=real_model,
```
Comment (BowenBao) [inline test/onnx/test_fx_to_onnx_with_onnxruntime.py:1094]: nit: Let's unify the arg name. It is called `exported_program` for input adapt, but `model` for output adapt.

File: torch/onnx/_internal/fx/passes/modularization.py [hunk 8]
```diff
@@ -214,12 +215,16 @@ def __init__(
...
             ]
         ],
+        is_exported_program: bool = False,
     ):
         self._module_stack = []
         if nn_module_stack_meta is None:
             return
         raw_meta = copy.copy(nn_module_stack_meta)
         for item in raw_meta.items():
+            if is_exported_program:
```
Comment (BowenBao) [inline torch/onnx/_internal/fx/passes/modularization.py:225]: not sure what this is about ..

File: torch/onnx/_internal/fx/passes/modularization.py [hunk 9]
```diff
@@ -826,7 +869,173 @@ def _run(self) -> torch.fx.GraphModule:
...
+        m._replace_hook(old=node, new=replace_with.name, user=user)
+
+    new_args = torch.fx.node.map_arg(user.args, maybe_replace_node)
+    new_kwargs = torch.fx.node.map_arg(user.kwargs, maybe_replace_node)
+    assert isinstance(new_args, tuple)
+    assert isinstance(new_kwargs, dict)
+    user.args = new_args
+    user.kwargs = new_kwargs
+
+    return
```
Comment (BowenBao) [inline torch/onnx/_internal/fx/passes/modularization.py:1041]: nit: remove

File: torch/onnx/_internal/fx/passes/modularization.py [hunk 10]
```diff
@@ -826,7 +869,173 @@ def _run(self) -> torch.fx.GraphModule:
...
+        to_module.register_parameter(target, from_obj)
+    elif attr_kind == fx_type_utils._AttrKind.BUFFER:
+        assert isinstance(from_obj, torch.Tensor)
+        to_module.register_buffer(target, from_obj, persistent=persistent)
+    elif attr_kind == fx_type_utils._AttrKind.CONSTANT:
+        assert isinstance(from_obj, (torch.Tensor, torch.ScriptObject))
+        setattr(to_module, target, from_obj)
+
+
+def _replace_the_use_with(
```
Comment (BowenBao) [inline torch/onnx/_internal/fx/passes/modularization.py:991]: Suggest a more specialized naming for this method, as well as the context and motivation in docstring.

My understanding is this is specifically for replacing `placeholder` with `getattr` inside each submodule, hence there are specific logic around dealing with multiple users, constructing stack meta, hooks etc.

File: torch/onnx/_internal/io_adapter.py (hunk 1 above)
Comment (thiagocrepaldi) [reply torch/onnx/_internal/io_adapter.py:69]: the non-exportedprogram also uses the input adapter. i dont think you can removed the torch.nn.Module from the type hint

The `Callable` type hiunt was discontinue by `torch.export` and I find a good idea to remove this even from the non-torchexported program.

File: torch/onnx/_internal/fx/type_utils.py [hunk 11]
```diff
@@ -25,6 +27,12 @@
     import onnx.defs.OpSchema.AttrType  # type: ignore[import]
 
 
+class _AttrKind(enum.Enum):
```
Comment (thiagocrepaldi) [inline torch/onnx/_internal/fx/type_utils.py:30]: torch.export defines similar enum. why not reuse it?

File: torch/onnx/_internal/fx/passes/modularization.py [hunk 12]
```diff
@@ -818,6 +828,39 @@ class Modularize(_pass.Transform):
...
+        module: torch.fx.GraphModule,
+        exported_program: Optional[torch.export.ExportedProgram] = None,
+    ):
+        super().__init__(diagnostic_context, module)
+        self.module = module
+        self.is_exported_program = exported_program is not None
+        self.exported_program = (
+            exported_program if exported_program is not None else None
+        )
+        self.graph_signature = (
```
Comment (BowenBao) [inline torch/onnx/_internal/fx/passes/modularization.py:844]: out of scope of this PR, but I wonder what is the cost to have something similar for non exported_program path?

File: torch/onnx/_internal/exporter.py [hunk 13]
```diff
@@ -687,9 +687,7 @@ def __init__(
     def __call__(
         self,
         *args: Any,
-        model_with_state_dict: Optional[
-            Union[torch.nn.Module, Callable, torch_export.ExportedProgram]
-        ] = None,
+        exported_program: Optional[torch_export.ExportedProgram] = None,
```
Comment (thiagocrepaldi) [inline torch/onnx/_internal/exporter.py:690]: we are not deprecating the non torchexported program just yet. we should keep the torch.nn.module type hint

File: torch/onnx/_internal/exporter.py [hunk 14]
```diff
@@ -698,26 +696,37 @@ def __call__(
         Args:
             args: The positional inputs to the model.
             kwargs: The keyword inputs to the model.
-            model_with_state_dict: The PyTorch model to fetch state from.
-                Required when :func:`enable_fake_mode` is used to extract real initializers as needed by the ONNX graph.
+            exported_program: Required when :func:`enable_fake_mode` is used to extract real initializers as
```
Comment (thiagocrepaldi) [inline torch/onnx/_internal/exporter.py:699]: how about the non-exportedprogram path?

**Review by thiagocrepaldi — COMMENTED** [review]
I didn't have time to look at the entire file PR, but as a general comment, if a change is not actually needed for modularization, let's create a separate PR and keep this change focused.

There are a couple changes that seems optional and should not be merged unless Modularization cannot work without it. In that case, let's discuss what are the blocking issues

* Change public facing API;
* Deprecating `torch.nn.Module` and Callable from the public facing APIs;
* Sinking initializers inside the model instead of keeping as inputs

File: test/onnx/test_fx_to_onnx_with_onnxruntime.py (hunk 7 above)
Comment (BowenBao) [inline test/onnx/test_fx_to_onnx_with_onnxruntime.py:1094]: I see the workaround you were talking about in the meeting.

IIUC, in order for fake exported model to have non-persistent buffers, we need to concretely torch.export.export the model, and stash the buffers back?

Comment (BowenBao) [reply test/onnx/test_fx_to_onnx_with_onnxruntime.py:1094]: How about we hijack the initialization of non-persistent buffer and record a concrete tensor under our export-purpose fake mode context manager when the model was being initialized? And non-persistent buffers initialized at runtime should get traced so we should have everything covered?

File: torch/onnx/_internal/io_adapter.py (hunk 1 above)
Comment (titaiwangms) [reply torch/onnx/_internal/io_adapter.py:69]: I deleted it because nn.Module is not using it in any cases. Specifically, this "model" args is only needed for exported program with fake mode. But if we want to keep it, we can keep it.

File: torch/onnx/_internal/fx/passes/modularization.py (hunk 8 above)
Comment (titaiwangms) [reply torch/onnx/_internal/fx/passes/modularization.py:225]: It's to skip the first nn module stack. In exported program, it also records another call stack into it. Added comment

File: torch/onnx/_internal/fx/passes/modularization.py [hunk 15]
```diff
@@ -826,7 +869,173 @@ def _run(self) -> torch.fx.GraphModule:
...
+                    user=user,
+                    propagate_meta=True,
+                )
+            graph.erase_node(node)
+
+        # reassign (recompile the graph_module)
+        graph_module.graph = graph
+
+    @_beartype.beartype
+    def _assign_state_dict_to_graph_module(
```
[the remark this answers is not in the document]
Comment (titaiwangms) [reply torch/onnx/_internal/fx/passes/modularization.py:924]: Part of

File: torch/onnx/_internal/exporter.py (hunk 14 above)
Comment (titaiwangms) [reply torch/onnx/_internal/exporter.py:699]: non-exportedprogram path doesn't need to pass in model to run.

File: torch/onnx/_internal/exporter.py [hunk 16]
```diff
@@ -698,26 +696,37 @@ def __call__(
...
+                fetch the initializers.
             options: The options to use for running the model with ONNX Runtime.
 
         Returns:
             The model output as computed by ONNX Runtime
         """
         import onnxruntime  # type: ignore[import]
 
-        # model specified by the user has precedence, when specified
-        model_with_state_dict = model_with_state_dict or self._model_torch
```
Comment (titaiwangms) [inline torch/onnx/_internal/exporter.py:711]: @thiagocrepaldi I will revert the refactor here, but there are something we can think about I think. What purpose is torch.nn.Module serving as a parameter in this __call__ and adapt_torch_inputs_to_onnx when it doesn't provide anything (it doesn't have stat_dict and mdel_signature). Look on the previous code, nn.Module is just fitting here to compromise ExportedProgram usage. That's why I proposed exported_program specific parameter to de-confused. I really thought nn.Module did something in this function.

File: torch/onnx/_internal/exporter.py (hunk 13 above)
Comment (titaiwangms) [reply torch/onnx/_internal/exporter.py:690]: Reverting it back. But wanted to remind, if users pass in nn.Module, it doesn't make any effect.

File: torch/onnx/_internal/exporter.py (hunk 4 above)
Comment (titaiwangms) [reply torch/onnx/_internal/exporter.py:920]: Oh it's inside  torch_export_graph_extractor that I use option.fake_context is not None to conditionally select the appended adapters.

File: torch/onnx/_internal/exporter.py (hunk 6 above)
Comment (titaiwangms) [reply torch/onnx/_internal/exporter.py:946]: Reverting this back

File: test/onnx/test_fx_to_onnx_with_onnxruntime.py (hunk 7 above)
Comment (titaiwangms) [reply test/onnx/test_fx_to_onnx_with_onnxruntime.py:1094]: @thiagocrepaldi What do you think? Serialize non-persistent buffers in fake mode context manager might be doable.

File: torch/onnx/_internal/exporter.py (hunk 4 above)
Comment (titaiwangms) [reply torch/onnx/_internal/exporter.py]: I didn't articulate the reason changing it. 

https://github.com/pytorch/pytorch/blob/91d1d2c421d0e94c8d3696dc27a44b466e34cf3e/torch/onnx/_internal/exporter.py#L711

I just wanted to make it clear that nn.Module doesn't need adapter because it is not prociding anything during adapting.

File: torch/onnx/_internal/fx/passes/modularization.py (hunk 12 above)
Comment (titaiwangms) [reply torch/onnx/_internal/fx/passes/modularization.py:847]: To construct graph_signature? Probably AOT export

Comment (thiagocrepaldi) [reply torch/onnx/_internal/fx/passes/modularization.py:847]: We could instantiate a graphsignature class and build it ourselves

**Comment by titaiwangms** [comment]
As offline meeting, will have another PR to keeping lifted inputs with modularization.
